spi-proposal.md

```diff
@@ -0,0 +1,354 @@
+
+### SPI Config Device
+
+Send once for each SPI device, specifying a new deviceId for each device
+(beginning from 0).
+
+#### bitOrder
+```
+LSBFIRST = 0
+MSBFIRST = 1
+```
+
+#### dataModes
+
+| mode    | clock polarity (CPOL) | clock phase (CPHA) |
+| --------|-----------------------|--------------------|
+| 0       | 0                     | 0                  |
+| 1       | 0                     | 1                  |
+| 2       | 1                     | 0                  |
+| 3       | 1                     | 1                  |
+
+
+#### optionalPins
+
+The `optionalPins` byte allows HW pins such as chip select, data ready, and
+other common (or perhaps even uncommon) actions realated to SPI HW that are
+not directly handled by SPI libraries. Optional pins are identified as bits of
+the `optionalPins` byte.
+
+```
+bit 0 = chip select
+bit 1 = data ready
+... additional bits TBD as needed
+```
+If one or more bits are set, then the pin number and an options byte are
+defined for each pin (so 2 bytes per pin) in order beginning with bit 0.
+
+The optional pins are described below:
+
+```
+Chip select (CS):
+0:  csPin                (0-127)
+1:  csOptions:
+      csActive (bit 0)   (0 = LOW -> HIGH, 1 = HIGH -> LOW, default = 0)
+      csToggle (bit 1)   (0 = toggle after last byte,
+                          1 = toggle between bytes, default = 0)
+
+Data ready:
+0:  dataReadyPin         (0-127)
+1:  dataReadyOptions:
+      dataReadyActive (bit 0) (0 = LOW, 1 = HIGH, default = 1)
+```
+
+```
+0:  START_SYSEX
+1:  SPI_DATA             (0x68)
+2:  SPI_CONFIG_DEVICE    (0x00)
+3:  deviceId | channel   (bits 2-6: deviceId, bits 0-1: channel (0-3))
+4:  bitOrder | dataMode  (bit 2: bitOrder, bits 0-1: dataMode (0-3))
+                          see mode column in table above
+5:  clockSpeed           (bits 0 - 6)
+6:  clockSpeed           (bits 7 - 14)
+7:  clockSpeed           (bits 15 - 21)
+8:  clockSpeed           (bits 22 - 28)
+9:  clockSpeed           (bits 29 - 32)
+10: optionalPins         (see description above)
+... defined optional pins (2 bytes each)
+N:  END_SYSEX (N = 10 + (num optional pins * 2))
+```
+
+### SPI Transfer
+
+**Read a bybe for each byte written.**
+
+Send an array of data to transfer to the SPI slave device, expecting a byte
+returned for each byte written. Respond with a `SPI_REPLY` message.
+
+Use the **sequence** byte to specify if this is a one time transfer or if it is part
+of a sequence of transfers, or an ongoing stream of transfers. Define as
+follows:
+
+```
+0 = single transfer (no sequence)
+1 = start sequence (1st part of sequence)
+2 = 2nd part of sequence
+3 = 3rd part of sequence, etc...
+... up to 100, after 100, wrap back to 1
+127 = end sequence
+```
+
+```
+0:  START_SYSEX
+1:  SPI_DATA             (0x68)
+2:  SPI_TRANSFER_REQUEST (0x01)
+3:  deviceId | channel   (bits 2-6: deviceId, bits 0-1: channel)
+4:  sequence             (see note on sequence above)
+5:  numBytes to read and write
```

Comment:
One option is to use the remaining bits of the 2nd data byte (since it takes 2 7-bit midi data bytes to make an 8 bit value, leaving 6 unused bits) to indicate whether or not a byte written expects a response. This would of course take extra clock cycles to parse out, but maybe it's only used in a SPI_TRANSFER_REQUEST message for the bytes passed in that message.